ZiLing/Midterm-Review.py

```diff
@@ -0,0 +1,481 @@
+
+Comparison Oprators 
+
+Operator   Name           Example
+==         Equal          x == y, 2 == 1 
+!=       Not Equal       x != 7, y != y
+>      Greater than       3 > 2, 4 > 5
+<       Less than         1 < 2, 5 < 5 
+>=   Greater or Equal to   1 >= 1, 1 >= 4
+<=   less or equal to      x <= 3 , 4 <= 4 
+
+You usually use them with if statement or while statement
+"""
+
+# You can use more than one comparison operators in the same line 
+for i in range (356,9,-3): 
+    if 325 < i > 35:
+        print (i)
+
+"""
+Logical Oprator 
+
+operator    Returns True if          Example
+and     both statements are true     x < 5 and x > 1 
+or      either statement is true     x < 5 or x > 1 
+not     the statement is false       not(x > 5)
+
+Identity Operator 
+
+operator    Returns True if          Example
+is      both variables are the same   5 is 5, x is 5
+is not  variables are not the same    x is not 4 
+
+Membership Operator 
+
+operator    Returns True if                 Example
+in       object is element in smt        i in mydict
+not in      object not in smth           i not in mydict     
+
+All the operators above return a boolean type value (True or False)
+"""
+
+# You can mix mash all the operators together as long as it makes sense
+for i in range (1,101):
+    if not(x%2 == 0 and x > 5 ): 
+        print (i)
+
+"""
+Conditional statements 
+
+Statement      name   
+if              if      
+elif         else if
+else           else
+while          while 
+"""
+
+"""
+Range
+
+range is a type that gives a range of intergers, like how 
+string is a sequence of characters, range is a sequence 
+of integers
+
+range(start,end,number between each)
+"""
+# This will not print out every value in range
+print(range(4))
+
+# This will
+for i in range(4):
+    print(i)
+
+#After talking about math and boolean, lets talk a little about string 
+
+"""
+String 
+
+String is a sequence of characters, or rather, a list of characters,
+meaning all functions that works with list will work with string 
+in the same way, for example 
+"""
+x = 'hello'
+e = x[1] 
+
+# len() function is used if you want the length of a list, in this case 
+# a string
+
+print(len('sorrynotsorry'))
+
+# slice is used to get par of the string 
+
+string = "string"
+rin = string[2:5]
+stri = string[:4]
+ing = string[3:]
+wholeString = string[:]
+# you can also just do
+st = 'string'[:2]
+
+# you will get an empty string if the range is 0 
+empty = "empty"[3:3]
+
+"""
+STRING IS IMMUTABLE
+
+meaning you cannot change any single character inside a string 
+
+Teaversals: ways to loop through the entire string, here are two traverals
+"""
+
+for letter in "string": 
+    print(letter)
+
+x = 0 
+while x < len('string'):
+    print('string'[x])
+    x += 1
+
+"""
+str.format{}
+
+This is a way used to format a string by inputting
+different types into a string, for example
+
+"this is a {} statement since the {}th value is {}".format(True,1,3.4)
```

Comment:
Use the f-string format. It's much easier to read.
```python
f"this is a {True} statement since the {1}th value is {3.4}"
```

There are also r-strings (r for raw) which are useful if you need to use slashes in your string. For example:
```python
r"yes\no"
```
Without the use of `r` that string would have been:
```
yes
o
```
because `\n` is the linebreak operator. You can also combine f and r strings use rf:
```python
rf"This string uses {2} methods defined for strings so that it will automatically convert a variable to a string if the conversion is possible and avoids backslash hell with statements like yes/no"
```